mission-control/blog/state-based-alerting/index.md

```diff
@@ -0,0 +1,500 @@
+</TerminalOutput>
+
+
+
+### Setting Up State-Based Alerting with Mission Control
+
+Mission Control can monitor these states and alert when they indicate problems. Let's create a check to monitor deployment rollout status.
+
+Create a new file named `deployment-state-check.yaml`:
+
+```yaml
+apiVersion: canaries.flanksource.com/v1
+kind: Canary
+metadata:
+  name: deployment-state-check
+spec:
+  interval: 30
+  kubernetes:
+    - name: check-deployment-rollout
+      description: "Monitor deployment rollout state"
+      resource:
+        apiVersion: apps/v1
+        kind: Deployment
+        name: nginx-deployment
+        namespace: default
+      results:
+      - name: Available
+        selector: $.status.conditions[?(@.type=="Available")].status
+        condition: Equal
+        error: "False"
+      - name: Progressing
+        selector: $.status.conditions[?(@.type=="Progressing")].status
+        condition: Equal
+        error: "False"
+      - name: ProgressingReason
+        selector: $.status.conditions[?(@.type=="Progressing")].reason
+        condition: Equal
+        error: "ProgressDeadlineExceeded"
+      - name: ErrorMessage
+        selector: $.status.conditions[?(@.type=="Progressing")].message
+        display: true
+```
+This Canary check:
+1. Runs every 30 seconds (`interval: 30`).
+2. Targets the `Deployment` named `nginx-deployment` in the `default` namespace.
+3. Defines results based on JSONPath selectors applied to the Deployment's status:
+    - Checks if the `Available` condition status is `False`.
+    - Checks if the `Progressing` condition status is `False`.
+    - Checks if the `Progressing` condition reason is `ProgressDeadlineExceeded`.
+    - Captures the `Progressing` condition message for display (`display: true`).
+    An alert is triggered if any condition marked with `error:` is met.
+
+Use `kubectl` to apply the Canary resource definition to your cluster:
+
+```bash
+kubectl apply -f deployment-state-check.yaml
+```
+
+This command registers your state-based check in Mission Control, which will now monitor your Deployment's state.
+
+### Simulating a Failed Deployment
+
+Create a problematic Deployment to see state-based alerting in action.
+
+Create a file named `failing-deployment.yaml`. This YAML defines a Deployment named `failing-deployment`:
+
+```yaml
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: failing-deployment
+  namespace: default
+spec:
+  replicas: 3
+  selector:
+    matchLabels:
+      app: failing-app
+  strategy:
+    rollingUpdate:
+      maxSurge: 1
+      maxUnavailable: 1
+    type: RollingUpdate
+  template:
+    metadata:
+      labels:
+        app: failing-app
+    spec:
+      containers:
+      - name: container
+        image: nginx:latest
+        resources:
+          limits:
+            memory: "10Mi"  # Intentionally too small
+          requests:
+            memory: "10Mi"
+        ports:
+        - containerPort: 80
+```
```

Comment:
We should have a screenshot here to show how canary checker tells the exact error along with a time based graph